discovery/digitalocean/digitalocean.go

```diff
@@ -0,0 +1,243 @@
+	doLabelSize        = doLabel + "size"
+	doLabelStatus      = doLabel + "status"
+	doLabelFeatures    = doLabel + "features"
+	doLabelTags        = doLabel + "tags"
+)
+
+// DefaultSDConfig is the default DigitalOcean SD configuration.
+var DefaultSDConfig = SDConfig{
+	Port:            80,
+	Separator:       ",",
+	RefreshInterval: model.Duration(60 * time.Second),
+}
+
+// SDConfig is the configuration for DO based service discovery.
+type SDConfig struct {
+	AccessToken     config_util.Secret `yaml:"access_token,omitempty"`
+	AccessTokenFile string             `yaml:"access_token_file,omitempty"`
+
+	// A tag that can be used to filter the droplets.
+	Tag string `yaml:"tag,omitempty"`
+
+	RefreshInterval model.Duration `yaml:"refresh_interval,omitempty"`
+	Port            int            `yaml:"port"`
+	Separator       string         `yaml:"separator,omitempty"`
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	if c.AccessToken == "" && c.AccessTokenFile == "" {
+		return fmt.Errorf("access_token or access_token_file is required")
+	}
+	if c.AccessToken != "" && c.AccessTokenFile != "" {
+		return fmt.Errorf("access_token and access_token_file are mutually exclusive")
+	}
+	return nil
+}
+
+// Discovery periodically performs DigitalOcean requests. It implements
+// the Discoverer interface.
+type Discovery struct {
+	*refresh.Discovery
+	client    *godo.Client
+	tag       string
+	port      int
+	separator string
+}
+
+// NewDiscovery returns a new Discovery which periodically refreshes its targets.
+func NewDiscovery(conf *SDConfig, logger log.Logger) (*Discovery, error) {
+	d := &Discovery{
+		tag:       conf.Tag,
+		port:      conf.Port,
+		separator: conf.Separator,
+	}
+
+	var rt http.RoundTripper = &http.Transport{
+		IdleConnTimeout: 5 * time.Duration(conf.RefreshInterval),
+		DialContext: conntrack.NewDialContextFunc(
+			conntrack.DialWithTracing(),
+			conntrack.DialWithName("digitalocean_sd"),
+		),
+	}
+	if conf.AccessToken != "" {
+		rt = config_util.NewBearerAuthRoundTripper(conf.AccessToken, rt)
+	}
+	if conf.AccessTokenFile != "" {
+		rt = config_util.NewBearerAuthFileRoundTripper(conf.AccessTokenFile, rt)
+	}
+
+	var err error
+	d.client, err = godo.New(
+		&http.Client{
+			Transport: rt,
+			Timeout:   5 * time.Duration(conf.RefreshInterval),
+		},
+		godo.SetUserAgent(fmt.Sprintf("Prometheus/%s", version.Version)),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error setting up digital ocean agent: %w", err)
+	}
+
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"digitalocean",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d, nil
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	source := "DO"
+	if d.tag != "" {
+		source += "_" + d.tag
```

Comment:
Why do you do this?

Reply:
Just as it made the tag made the source unique. Will remove as we will remove tag anyway.

Reply:
This is still here.